reactor-netty-core/src/test/java/reactor/netty/resources/PooledConnectionProviderCustomMetricsTest.java

```diff
@@ -69,36 +70,47 @@ void tearDown() throws Exception {
 
 	@Test
 	void customRegistrarIsUsed() {
-		AtomicBoolean used = new AtomicBoolean();
+		AtomicBoolean registered = new AtomicBoolean();
+		AtomicBoolean deRegistered = new AtomicBoolean();
 
-		triggerAcquisition(true, () -> (a, b, c, d) -> used.set(true));
-		assertThat(used.get()).isTrue();
+		triggerAcquisition(true, () -> new MeterRegistrarImpl(registered, deRegistered, null));
+		assertThat(registered.get()).isTrue();
+		assertThat(deRegistered.get()).isFalse();
+
+		pool.dispose();
+		assertThat(deRegistered.get()).isTrue();
 	}
 
 	@Test
 	void connectionPoolMaxMetrics() {
-		AtomicInteger maxAllocSize = new AtomicInteger();
-		AtomicInteger maxPendingAcquireSize = new AtomicInteger();
-		triggerAcquisition(true, () -> (a, b, c, d) -> {
-			maxAllocSize.set(d.maxAllocatedSize());
-			maxPendingAcquireSize.set(d.maxPendingAcquireSize());
-		});
-		assertThat(maxAllocSize.get()).isEqualTo(MAX_ALLOC_SIZE);
-		assertThat(maxPendingAcquireSize.get()).isEqualTo(MAX_PENDING_ACQUIRE_SIZE);
+		AtomicInteger customMetric = new AtomicInteger();
+
+		triggerAcquisition(true, () -> new MeterRegistrarImpl(null, null, customMetric));
+		assertThat(customMetric.get()).isEqualTo(MAX_ALLOC_SIZE);
 	}
 
 	@Test
 	void customRegistrarSupplierNotInvokedWhenMetricsDisabled() {
-		AtomicBoolean used = new AtomicBoolean();
+		AtomicBoolean registered = new AtomicBoolean();
 
-		triggerAcquisition(false, () -> {
-			used.set(true);
-			return null;
-		});
-		assertThat(used.get()).isFalse();
+		triggerAcquisition(false, () -> new MeterRegistrarImpl(registered, null, null));
+		assertThat(registered.get()).isFalse();
+	}
+
+	@Test
+	void disposeWhenMetricsDeregistered() {
+		AtomicBoolean registered = new AtomicBoolean();
+		AtomicBoolean deRegistered = new AtomicBoolean();
+
+		Connection conn = triggerAcquisition(true, () -> new MeterRegistrarImpl(registered, deRegistered, null));
+		assertThat(registered.get()).isTrue();
+		assertThat(deRegistered.get()).isFalse();
+
+		pool.disposeWhen(remoteAddress.get());
+		assertThat(deRegistered.get()).isTrue();
 	}
 
-	private void triggerAcquisition(boolean metricsEnabled, Supplier<ConnectionProvider.MeterRegistrar> registrarSupplier) {
+	private Connection triggerAcquisition(boolean metricsEnabled, Supplier<ConnectionProvider.MeterRegistrar> registrarSupplier) {
```

Comment:
Connection establishment is expected to fail always so we cannot return any Connection. The intention of this method is just to create the connection pool.

```
io.netty.channel.AbstractChannel$AnnotatedSocketException: Can't assign requested address: localhost/127.0.0.1:0
Caused by: java.net.BindException: Can't assign requested address
	at sun.nio.ch.Net.connect0(Native Method)
	at sun.nio.ch.Net.connect(Net.java:482)
	at sun.nio.ch.Net.connect(Net.java:474)
```

```suggestion
	private void triggerAcquisition(boolean metricsEnabled, Supplier<ConnectionProvider.MeterRegistrar> registrarSupplier) {
```